src/experiments/logins/api.js

```diff
@@ -0,0 +1,210 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/* globals AppConstants, browser, Components, CustomizableUI, dispatcher,
+   ExtensionCommon, ExtensionAPI, Services, XPCOMUtils */
+
+"use strict";
+
+ChromeUtils.defineModuleGetter(this, "ExtensionUtils",
+                               "resource://gre/modules/ExtensionUtils.jsm");
+ChromeUtils.defineModuleGetter(this, "LoginHelper",
+                               "resource://gre/modules/LoginHelper.jsm");
+ChromeUtils.defineModuleGetter(this, "Services",
+                               "resource://gre/modules/Services.jsm");
+
+const { ExtensionError } = ExtensionUtils;
+
+const getLogins = () => {
+  const logins = Services.logins.
+    getAllLogins().
+    filter((l) => !(l.hostname || "").startsWith("chrome://")).
+    map(LoginHelper.loginToVanillaObject);
+  return logins;
+};
+
+const getLogin = (id) => {
+  const login = getLogins().
+    filter(l => l.guid === id)[0];
+  return login || null;
+};
+
+this.logins = class extends ExtensionAPI {
+  getAPI(context) {
+    const EventManager = ExtensionCommon.EventManager;
+    return {
+      experiments: {
+        logins: {
+          // See schema.json for function documentation.
+          // The Firefox internal code throws if the wrong type is passed into
+          // the LoginManager methods. To reduce the odds of those mistakes,
+          // this code follows the following naming convention:
+          //   - fooLogin = login as vanilla JS object
+          //   - fooLoginInfo = login as nsILoginInfo
+          //   - fooBag = (subset of) login fields as nsIPropertyBag
+          getAll() {
+            const logins = getLogins();
+            return logins;
+          },
+          get(id) {
+            const login = getLogin(id);
+            return login;
+          },
+          add(login) {
+            if (getLogin(login.guid)) {
+              throw new ExtensionError(`Add failed: Login already exists with ID ${login.guid}`);
+            }
+            try {
+              const loginInfo = LoginHelper.vanillaObjectToLogin(login);
+              Services.logins.addLogin(loginInfo);
+              const addedLogin = getLogin(login.guid);
+              return addedLogin;
+            } catch (ex) {
+              throw new ExtensionError(ex);
+            }
+          },
+          update(newLogin) {
+            if (!("guid" in newLogin)) {
+              throw new ExtensionError(`Update failed: ID is required`);
+            }
+            // The addons framework code always supplies null values for missing keys,
+            // so filter those out.
+            // TODO: Could this cause subtle bugs? Would anyone ever want to update a field to be null? If so, maybe default to 'NULL" (the string), or undefined.
```

Comment:
This is my one remaining point of uncertainty. Feedback welcome

Reply:
I did not realize this is the case.  This `null` behavior seems counter to "common idioms" with JSON-ish usage pattern (using `null` to clear a value, and `undefined` or omitted to ignore), arguably best embodied in [JSON merge patch](https://tools.ietf.org/html/rfc7386).

I would lean toward `newLogin` then having to be the complete representation rather than the subset that should actually be updated.

Reply:
To be  clear, I don't _think_ this behavior will have any impact on us in practice --- for now.  If we ever try to tackle toggling between "HTTP auth" and "submit form", this will cause problems.